### ISSUE TITLE
Migration to Rocky Linux removes symbolic links in /etc/alternatives for java 

### ISSUE THREAD

**Issue by klaus-tux** [rocky-linux/rocky-tools#41]
I just migrated a CentOS 8 system to Rocky Linux that has the following java packages installed:
```
java-1.8.0-openjdk.x86_64                   1:1.8.0.292.b10-1.el8_4                @appstream   
java-1.8.0-openjdk-headless.x86_64          1:1.8.0.292.b10-1.el8_4                @appstream   
java-11-openjdk.x86_64                      1:11.0.11.0.9-2.el8_4                  @appstream   
java-11-openjdk-devel.x86_64                1:11.0.11.0.9-2.el8_4                  @appstream   
java-11-openjdk-headless.x86_64             1:11.0.11.0.9-2.el8_4                  @appstream   
```
After the migration there are no symbolic links for all the java related tools in `/etc/alternatives` any more.
It would be great if you could improve the migration script to keep the symbolic links :-)

**Comment by msmucr** [comment]
I've also experienced this issue, however after some examination I found, it is not a fault of migration script. 
The script calls dnf to make a distro-sync, which essentially reinstall all the packages. It seems, upstream OpenJDK packages has some issue with corresponding entries among alternatives. I've tested, alternative entries doesn't survive dnf reinstall even at other distributions without any repository mangling or version change. 
Also it doesn't seem to be some general issue and it's really rather specific to OpenJDK, because other packages (like Python, cifs-utils etc.) doesn't have any problem with forced reinstall and keeping alternatives.

Of course @pajamian probably likely can extend the script functionality to tar /etc/alternatives and /var/lib/alternatives before actual upgrade and restore it after that. However to be honest, I don't think, such forced and optimistic approach is a good one. Some package installation can fail during the migration or there can be some other issue and you'll end up with restored invalid entries among alternatives (dead symlinks etc.). 

I haven't really debugged RPM spec files or why those original OpenJDK packages have this reinstall problem, but so far alternatives from those packages can be restored in several ways.
- by two step complete uninstall and install, but it could break other packages dependent on JDK or JRE
- manually vi alternatives (lots of entries with error prone typing)
- by manually running post-install scriptlet from original packages

IMO the third approach is the best. You can get post-install scriptlet by extracting from rpm -q package --scripts. There are commands for registration into alternatives.
For my automation purposes I've cooked following quick-n-dirty fix script bellow. Only -headless and -devel packages are actually necessary, just those registers alternatives like java, javac etc. 
When it finds some of listed package, it just run its post-install scriptlet in separate bash.

```
#!/bin/bash
if [[ $EUID -ne 0 ]]; then
    echo "This script must be run as root" 
    exit 1
fi

packages="java-1.8.0-openjdk-headless \
java-1.8.0-openjdk-devel \
java-11-openjdk-headless \
java-11-openjdk-devel"

for p in $packages
do
    rpm --quiet -q $p && \
        bash -c "$(rpm -q $p --scripts | sed -n '/postinstall/, /postuninstall/{ /postinstall/! { /postuninstall/ ! p} }')"
done

```
Of course, it just recreates default package alternatives, if you had some custom manual settings there, you'd need to set it again.

**Comment by pajamian** [comment]
I'll look into this, but it's going to be a bug in the packaging of java.  The appropriate remedy is to file a bug report upstream.  We can put a note in the issues section of the README so taht people are aware of it and of what the appropriate steps are to fix it once the migration is finished but I am reluctant to alter the script itself for this.

**Comment by klaus-tux** [comment]
Hi @msmucr !
Thank you very much for your thorough analysis of the problem. I fully agree, that the third approach is the best. I successfully applied your script :-)

@ @pajamian : I fully agree to not altering the script for this issue. But it would be nice to have a note in the README, so people don't open additional issues upon this. What do you think about adding the script from @msmucr to the repository?

**Comment by pajamian** [comment]
I reproduced the issue in RHEL and filed a bugzilla:
https://bugzilla.redhat.com/show_bug.cgi?id=1976053

**Comment by pajamian** [comment]
I've simplified the above script to a one-liner:
`rpm -qa --scripts java-{1.8.0,11}-openjdk-{headless,devel} | sed -n '/postinstall/, /exit/{ /postinstall/! { /exit/ ! p} }' | sh`

...tested and works

**Comment by msmucr** [comment]
Thank you both @pajamian and @klaus-tux. 
Adding the note to readme is definitely useful, until original upstream package will be fixed. The one-liner is definitely much more elegant ;), I've just done something quickly for batch fixing after migrations. Thanks for the report to RH bugzilla, I also planned to check those packages and report the issue during next week, so it's great, you've already done that.
